### PR TITLE
[C++] Support more pulsar command name and version in Wireshark dissector

### PR DESCRIPTION
**Pull request by Demogorgon314**
### Motivation

Currently, the Pulsar Wireshark dissector missing some support of the new command name, version, and server errors, we should support it.

### Modifications

1. Support missing pulsar command name.
2. Support missing pulsar server errors.
3. Support missing protocol version.
4. Some refactor

### Documentation

Need to update docs? 

- [x] `no-need-doc`

### CODE REVIEW

**Review by BewareMyPower — COMMENTED** [review]
Could you avoid unnecessary changes? I see a lot of code reformat here. I agree that it's better to reformat this file.

It looks like this file's format was not covered by `clang-format`, could you add the format to the script or `ci-cpp.yaml`? The current format check is done by `build-support/run_clang_format.py`.

File: pulsar-client-cpp/wireshark/pulsarDissector.cc [hunk 1]
```diff
@@ -78,62 +78,139 @@ static int hf_pulsar_message_permits = -1;
 
 static int ett_pulsar = -1;
 
-const static int FRAME_SIZE_LEN = 4;
+const static gint FRAME_SIZE_LEN = 4;
```
Comment (BewareMyPower) [inline pulsar-client-cpp/wireshark/pulsarDissector.cc:81]: Why do you change `int` to `gint`?

File: pulsar-client-cpp/wireshark/pulsarDissector.cc [hunk 2]
```diff
@@ -153,18 +230,15 @@ struct MessageIdComparator {
 
 struct RequestData {
     uint32_t requestFrame;
-    nstime_t requestTimestamp;
+    nstime_t requestTimestamp{};
```
Comment (BewareMyPower) [inline pulsar-client-cpp/wireshark/pulsarDissector.cc:233]: Why do you add `{}` here? The same question applies for other places and I won't comment again.

File: pulsar-client-cpp/wireshark/pulsarDissector.cc [hunk 3]
```diff
@@ -203,7 +277,7 @@ static void dissect_message_metadata(proto_tree* frame_tree, tvbuff_t *tvb, int
     }
 
     static MessageMetadata msgMetadata;
-    uint8_t* ptr = (uint8_t*) tvb_get_ptr(tvb, offset, metadataSize);
+    auto* ptr = (uint8_t*) tvb_get_ptr(tvb, offset, metadataSize);
```
Comment (BewareMyPower) [inline pulsar-client-cpp/wireshark/pulsarDissector.cc:280]: Just use `auto` instead of `auto*` if you don't want to declare type explicitly (because there is a conversion in right hand), the type deduction will work well.

File: pulsar-client-cpp/wireshark/pulsarDissector.cc [hunk 4]
```diff
@@ -445,10 +515,10 @@ static int dissect_pulsar_message(tvbuff_t *tvb, packet_info* pinfo, proto_tree*
...
         case BaseCommand::SEND: {
             const CommandSend& send = command.send();
-            RequestData& data = state->producers[send.producer_id()].messages[send.sequence_id()];
-            data.requestFrame = pinfo->fd->num;
-            data.requestTimestamp.secs = pinfo->fd->abs_ts.secs;
-            data.requestTimestamp.nsecs = pinfo->fd->abs_ts.nsecs;
+            RequestData& request_data = state->producers[send.producer_id()].messages[send.sequence_id()];
+            request_data.requestFrame = pinfo->fd->num;
+            request_data.requestTimestamp.secs = pinfo->fd->abs_ts.secs;
+            request_data.requestTimestamp.nsecs = pinfo->fd->abs_ts.nsecs;
```
Comment (BewareMyPower) [inline pulsar-client-cpp/wireshark/pulsarDissector.cc:521]: There's no need to rename `data` to `request_data`.

File: pulsar-client-cpp/wireshark/pulsarDissector.cc (hunk 1 above)
Comment (Demogorgon314) [reply pulsar-client-cpp/wireshark/pulsarDissector.cc:81]: You can see `tvb_get_ntohl(tvb, offset)`, the offset type is `gint`.

File: pulsar-client-cpp/wireshark/pulsarDissector.cc (hunk 4 above)
Comment (Demogorgon314) [reply pulsar-client-cpp/wireshark/pulsarDissector.cc:521]:
```c++
static int dissect_pulsar_message(tvbuff_t *tvb, packet_info* pinfo, proto_tree* tree, void* data _U_)
```
Here has a `data` too, it will make this can be confusing.

File: pulsar-client-cpp/wireshark/pulsarDissector.cc (hunk 3 above)
Comment (Demogorgon314) [reply pulsar-client-cpp/wireshark/pulsarDissector.cc]: Addressed.

**Comment by Demogorgon314** [comment]
> It looks like this file's format was not covered by `clang-format`, could you add the format to the script or `ci-cpp.yaml`? The current format check is done by `build-support/run_clang_format.py`.

Maybe I can open another PR to fix it.

File: pulsar-client-cpp/wireshark/pulsarDissector.cc (hunk 2 above)
Comment (Demogorgon314) [reply pulsar-client-cpp/wireshark/pulsarDissector.cc]: Removed.

File: pulsar-client-cpp/wireshark/pulsarDissector.cc (hunk 1 above)
Comment (BewareMyPower) [reply pulsar-client-cpp/wireshark/pulsarDissector.cc:79]: If you insisted on the type safety, you should notice that `FRAME_SIZE_LEN` is also used in

```c++
uint32_t offset = FRAME_SIZE_LEN;
```

and

```c++
static uint32_t get_pulsar_message_len(packet_info *pinfo _U_, tvbuff_t *tvb, int offset,
                                    void *data _U_) {
    uint32_t len = (uint32_t) tvb_get_ntohl(tvb, offset);
    return FRAME_SIZE_LEN + len;
}
```

The type of `FRAME_SIZE_LEN` should be `uint32_t` actually.

Instead of changing the type, it's better to add the type cast explicitly to declare the type cast is safe. For example,

```c++
    uint32_t cmdSize = (uint32_t) tvb_get_ntohl(tvb, static_cast<gint>(offset));
```

However, it's not a big issue if you keep the current code. It's just something about the code quality.

File: pulsar-client-cpp/wireshark/pulsarDissector.cc [hunk 5]
```diff
@@ -295,22 +367,20 @@ void link_to_response_frame(proto_tree* cmd_tree, tvbuff_t* tvb, int offset, int
...
     }
 
-    uint32_t offset = FRAME_SIZE_LEN;
-    int maxOffset = tvb_captured_length(tvb);
+    gint offset = FRAME_SIZE_LEN;
+    guint maxOffset = tvb_captured_length(tvb);
 
-    uint32_t cmdSize = (uint32_t) tvb_get_ntohl(tvb, offset);
+    guint32 cmdSize = tvb_get_ntohl(tvb, offset);
     offset += 4;
```
Comment (BewareMyPower) [inline pulsar-client-cpp/wireshark/pulsarDissector.cc:384]: The same as I've said before. The type casting between `quint32` and `uint32_t` is safe. There is no need to modify the original code. Because for the same reason, if someone else see the following code

```c++
   // offset is uint32_t, while cmdSize is guint32 after your change
   offset += cmdSize;
``` 

He might also want to change the code back or add the casting like

```c++
offset += static_cast<uint32_t>(cmdSize);
```

Therefore, stop making these unnecessary changes.

File: pulsar-client-cpp/wireshark/pulsarDissector.cc (hunk 4 above)
Comment (BewareMyPower) [reply pulsar-client-cpp/wireshark/pulsarDissector.cc:519]: The `data` variable is passed to

```c++
link_to_response_frame(cmd_tree, tvb, cmdOffset, cmdSize, data);
```

```c++
void link_to_response_frame(proto_tree* cmd_tree, tvbuff_t* tvb, int offset, int size,
                            const RequestData& reqData) {
```

I don't think there's something confusing.

Comment (BewareMyPower) [reply pulsar-client-cpp/wireshark/pulsarDissector.cc:519]: And if you want to keep the same coding style, you should have noticed in the other `case` blocks, the variable name `reqData` is used. However, as a local variable whose lifetime is just inside the `case` block, the name `data` is just okay.

File: pulsar-client-cpp/wireshark/pulsarDissector.cc [hunk 6]
```diff
@@ -795,7 +865,7 @@ static int dissect_pulsar_message(tvbuff_t *tvb, packet_info* pinfo, proto_tree*
                 proto_tree_add_string(cmd_tree, hf_pulsar_producer_name, tvb, cmdOffset, cmdSize,
                                       success.producer_name().c_str());
 
-                proto_tree* item = proto_tree_add_uint64(cmd_tree, hf_pulsar_producer_id, tvb,
+                item = proto_tree_add_uint64(cmd_tree, hf_pulsar_producer_id, tvb,
```
Comment (BewareMyPower) [inline pulsar-client-cpp/wireshark/pulsarDissector.cc:868]: C++ allows sharing the same variable name inside and outside the `if` / `for` block. Like

```c++
    int x = 0;
    if (x == 0) {
        int x = 100;
        std::cout << x << std::endl;
    }
```

The same variable in the code block will shadow the variable outside the code block.

The original code is better because it avoids accessing the global variable, whose definition is somewhere far and might be modified.

File: pulsar-client-cpp/wireshark/pulsarDissector.cc [hunk 7]
```diff
@@ -1025,7 +1033,7 @@ extern __attribute__((unused)) WS_DLL_PUBLIC_DEF const int plugin_want_minor = V
 WS_DLL_PUBLIC void plugin_register(void);
 
 __attribute__((unused)) static void
-proto_reg_handoff_pulsar(void)
+proto_reg_handoff_pulsar()
```
Comment (BewareMyPower) [inline pulsar-client-cpp/wireshark/pulsarDissector.cc:1036]: There is something special for the `void` argument in C++. See https://softwareengineering.stackexchange.com/questions/286490/what-is-the-difference-between-function-and-functionvoid.

Though this change doesn't affect, you can see the definition of `proto_plugin` here: https://www.wireshark.org/docs/wsar_html/structproto__plugin.html

The function pointer's type is

```c++
void(* register_handoff )(void)
```

So please just keep the code not changed.

**Comment by BewareMyPower** [comment]
I must repeat the suggestion that please avoid unnecessary changes about the code style. Some are just about the code style. The code style of this file is more C-style though it still uses C++'s semantics.

If you want it to be more C++ style, there could be many more changes.

File: pulsar-client-cpp/wireshark/pulsarDissector.cc [hunk 8]
```diff
@@ -157,10 +232,7 @@ struct RequestData {
     uint32_t ackFrame;
     nstime_t ackTimestamp;
 
-    RequestData()
-            : requestFrame(UINT32_MAX),
-              ackFrame(UINT32_MAX) {
-    }
+    RequestData(): requestFrame(UINT32_MAX),ackFrame(UINT32_MAX) {}
```
Comment (BewareMyPower) [inline pulsar-client-cpp/wireshark/pulsarDissector.cc:235]: In addition, I see many changes about code format, like here. The change is really redundant. For example, when I use `clang-format` to reformat your file again, this line also doesn't meet the code format.

```diff
-    RequestData(): requestFrame(UINT32_MAX),ackFrame(UINT32_MAX) {}
+    RequestData() : requestFrame(UINT32_MAX), ackFrame(UINT32_MAX) {}
```

Comment (BewareMyPower) [reply pulsar-client-cpp/wireshark/pulsarDissector.cc:235]: Please just retain the necessary changes. Regarding to the code style and format, we can use another PR to fix it.

Comment (Demogorgon314) [reply pulsar-client-cpp/wireshark/pulsarDissector.cc:235]: Make sense. I'll redo this change.

File: pulsar-client-cpp/wireshark/pulsarDissector.cc (hunk 4 above)
Comment (Demogorgon314) [reply pulsar-client-cpp/wireshark/pulsarDissector.cc:519]: I change to `reqData ` now.

File: pulsar-client-cpp/wireshark/pulsarDissector.cc (hunk 6 above)
Comment (Demogorgon314) [reply pulsar-client-cpp/wireshark/pulsarDissector.cc:868]: Right, the original code has a variable defined here, it never uses it.
```c++
    proto_tree* item = NULL;

    switch (command.type()) {
```

File: pulsar-client-cpp/wireshark/pulsarDissector.cc (hunk 7 above)
Comment (Demogorgon314) [reply pulsar-client-cpp/wireshark/pulsarDissector.cc:1036]: Thanks for your remind, Addressed.

File: pulsar-client-cpp/wireshark/pulsarDissector.cc (hunk 5 above)
Comment (Demogorgon314) [reply pulsar-client-cpp/wireshark/pulsarDissector.cc]: Addressed.

File: pulsar-client-cpp/wireshark/pulsarDissector.cc (hunk 1 above)
Comment (Demogorgon314) [reply pulsar-client-cpp/wireshark/pulsarDissector.cc]: Thanks for your reminder, I changed it back.

File: pulsar-client-cpp/wireshark/pulsarDissector.cc (hunk 6 above)
Comment (BewareMyPower) [reply pulsar-client-cpp/wireshark/pulsarDissector.cc]: Yeah, we can open another PR to fix the code style.

**Review by BewareMyPower — APPROVED** [review]

**Review by codelipenghui — APPROVED** [review]
